assets/javascripts/kitten/hoc/media-queries.js

```diff
@@ -29,6 +29,16 @@ export const mediaQueries = (WrappedComponent, hocProps = {}) =>
       }, {})
     }
 
+    cloneMethods = wrappedComponentInstance => {
```

Comment:
### Concernant le passage de méthode d'API de composants wrappés par un HOC

Après quelques recherches dans les internets, la solution qui semble être la plus adéquate dans notre cas, ce serait : https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components. TLDR : Cette technique permet d'accéder à des `ref` enfants, que ce soient des noeuds DOM ou des éléments React.

Et là, on se dit "Cool ! Allons-y !". 

:fire: :fire:  **TOO BAD, you were too happy, I needed to stop your enthusiasm**. :fire: :fire: 

Oui, donc c'est cette technique qu'on devrait utiliser mais cela ne fonctionne pas actuellement avec Redux : https://github.com/reduxjs/react-redux/issues/914. C'est apparement dans leur roadmap pour la prochaine release : https://github.com/reduxjs/react-redux/issues/950 (j'ai pas du tout eu la force de lire tout le thread). 

Donc, dans un premier temps, je propose de garder la solution de @LanF3usT qui est astucieuse en la modifiant légèrement pour éviter tout effet de bord. Je n'ai jamais vraiment creusé quelles propriétés sont exposées sur une instance de composant, donc j'aimerais qu'on évite de les exposer sur le HOC de façon inutile (surtout que ce HOC est utilisé dans de nombreux composants). 

Le souci relevé ici n'est encore jamais apparu car seul `Modal` et `CartRewardCard` ont des méthodes de type API (qui permettent d'agir sur le composant de l'extérieur).

Proposition discutée avec @LanF3usT : 
```js
// Sur le composant wrappé
export const CartRewardCard = mediaQueries(CartRewardCardComponent, {
  viewportIsMobile: true,
  exposedMethods: ['close'],
})

// Sur le HOC mediaQueries
export const mediaQueries = (WrappedComponent, hocProps = {}) =>
  class MediaQueriesWrapper extends Component {
    …

    setExposedMethods = (wrappedComponentInstance) => {
      if (!wrappedComponentInstance) return
      if (!hocProps.exposedMethods) return

      hocProps.exposedMethods.forEach(method => {
        this[method] = wrappedComponentInstance[method]
      })
    }

    …

    render() {
      return (
        <WrappedComponent
          ref={this.setExposedMethods}
          {...this.props}
          {...this.state}
        />
      )
    }
  }
```

Reply:
Trop dommage les `forwardRef`, c'est exactement ce qu'on voulait :'(

Go pour `exposedMethods`, ça sera plus sûr !

Reply:
Ouais :/

Je viens de tester avec le `exposedMethods`. Ca fonctionne bien et permet de limiter les méthodes exposées sauf que la déclaration se fait à des endroits non souhaités. Dans mon exemple de cart, je suis obligé d'ajouter `exposedMethods: ['close']` dans le composant `CartRewardCard` sauf qu'à cet endroit là, je ne devrait pas avoir connaissance de la modale. Ca doit être à l'app de faire ce choix de composition. J'ai l'impression que l'exposition automatique est plus simple finalement et répond mieux à ce qu'on souhaite faire. En attendant `forwardRef` qui fera la même chose. Vous en pensez quoi ?

Reply:
Mon commentaire au dessus n'est pas valable. J'avais en tête le `close` de la modale alors que c'est bien le `close` de `CartRewardCard` qu'on souhaite exposer.